[clang][analyzer] Bring cplusplus.ArrayDelete out of alpha (#83985)
llvm/llvm-project: clang/docs/analyzer/checkers.rst

C++ Checkers.

.. _cplusplus-ArrayDelete:

cplusplus.ArrayDelete (C++)

> """""""""""""""""""""""""""
> Reports destructions of arrays of polymorphic objects that are destructed as

Reviewer: ```suggestion
"""""""""""""""""""""""""""

Reports destructions of arrays of polymorphic objects that are destructed as
```

Style nit.

> their base class. If the dynamic type of the array's object is different from
> its static type, calling `delete[]` is undefined.

Reviewer: ```suggestion
their base class. If the dynamic type of the array is different from
its static type, calling `delete[]` is undefined.
```

> This checker corresponds to the CERT rule `EXP51-CPP: Do not delete an array through a pointer of the incorrect type <https://wiki.sei.cmu.edu/confluence/display/cplusplus/EXP51-CPP.+Do+not+delete+an+array+through+a+pointer+of+the+incorrect+type>`_.

Reviewer: ```suggestion
This checker corresponds to the SEI CERT rule `EXP51-CPP: Do not delete an array through a pointer of the incorrect type <https://wiki.sei.cmu.edu/confluence/display/cplusplus/EXP51-CPP.+Do+not+delete+an+array+through+a+pointer+of+the+incorrect+type>`_.
```

Style nit, to toe the line.

.. code-block:: cpp

 class Base {
 public:
   virtual ~Base() {}
 };
 class Derived : public Base {};

 Base *create() {
   Base *x = new Derived[10]; // note: Casting from 'Derived' to 'Base' here
   return x;
 }

 void foo() {
   Base *x = create();
   delete[] x; // warn: Deleting an array of 'Derived' objects as their base class 'Base' is undefined
 }

> **Limitations**

Reviewer: (Shouldn't this be a sub-heading?)

Reviewer 2: No, this is the formatting that's consistently used in this document.
